app/src/lib/stores/app-store.ts

```diff
@@ -3136,6 +3136,20 @@ export class AppStore extends TypedBaseStore<IAppState> {
 
     await this.updateBranchProtectionsFromAPI(repository)
 
+    const gitStore = this.gitStoreCache.get(repository)
+
+    if (
+      gitStore.currentRemote &&
+      repository.gitHubRepository &&
+      repository.gitHubRepository.cloneURL !== apiRepo.clone_url
+    ) {
+      await this._setRemoteURL(
+        repository,
+        gitStore.currentRemote.name,
+        apiRepo.clone_url
+      )
+    }
+
```

Comment:
Hey @niik and @outofambit, this is @kuychaco and my attempt to fix the bug described in this comment. https://github.com/desktop/desktop/issues/3855#issuecomment-522170778. It appears like what we have fixes that bug, but before we get deeper into testing and validating it we wanted to get a sanity check to the approach. 

1. Does it make sense to update the origin url in `.git/config` (using setRemoteUrl) when the remote url changes on the GitHub side?
2. Are there instances when using `gitStore.currentRemote.name` will cause problems? Should we use `gitStore.defaultRemote.name` instead?
3. Assuming we should update the remote url, Is this the right place to update the remote url?

Reply:
@kuychaco @probablycorey this looks like a good approach to me! 

1. yep! that's ideal
2. hmm i'm not sure we want to update non-default github remote urls... there might be multiple githubby remotes for a repo (eg `origin` and `upstream`). i think we have to make sure `gitStore.defaultRemote.url` is the same as `repository.gitHubRepository.cloneURL`, and update the url as you have here.
3. i think this is the right place to do it!
4. we might need to call [`gitStore.loadRemotes()`](https://github.com/desktop/desktop/blob/f96b6930a420887923db6e2b9b008a444f515c52/app/src/lib/stores/git-store.ts#L1075-L1099) after we the call to `_setRemoteUrl` to make sure other references to that remote in memory are updated to the new url. (maybe that call should be inside of `_setRemoteUrl`?)

Reply:
i also updated the docs for `repositoryWithRefreshedGitHubRepository` while i was looking into this. going to push that into this PR if that's ok!

Reply:
One thing that concerns me with this approach is that we'd potentially change the remote protocol as part of this. On GitHub Enterprise administrators are able to globally enforce that all repository `clone_url` use the ssh protocol and given how suboptimal our support for ssh is at the moment that makes me nervous. Vice versa some of our users are adamant about using the SSH protocol (and accept that they have to do some work of their own to make it work) and we could potentially end up switching them back to http.

> i think we have to make sure gitStore.defaultRemote.url is the same as repository.gitHubRepository.cloneURL

:+1: This has the added benefit of us never overwriting the remote URL if the user has intentionally changed it from the default.

Reply:
maybe we can frankenstein the URL path from the API onto the protocol+hostname already configured in the existing remote? this would preserve any user configured protocol or enforced ones from an admin. i'm pretty sure we don't need to worry about hostnames changing. from looking through our githubrepository store, it looks like if the enterprise hostname were to change, it wouldn't be updated anywhere in the desktop databases.

Reply:
we abandoned this approach after @kuychaco found that the `git@github.com` ssh urls are are not parseable by `URL.parse`.